File: server/utils/rate-limiter.ts
interface RateLimit {
  count: number;
  resetTime: number;
}

interface RateLimitConfig {
  maxRequests: number;
  windowMs: number;
}

// In-memory rate limiting (in production, use Redis)
const rateLimits = new Map<string, RateLimit>();

const RATE_LIMIT_CONFIGS: Record<string, RateLimitConfig> = {
  'DOCUMENT_VALIDATION': { maxRequests: 10, windowMs: 5 * 60 * 1000 }, // 10 per 5 minutes
  'REGISTRATION': { maxRequests: 3, windowMs: 60 * 60 * 1000 }, // 3 per hour
  'OCR_PROCESSING': { maxRequests: 5, windowMs: 10 * 60 * 1000 }, // 5 per 10 minutes
  'EMAIL_VALIDATION': { maxRequests: 20, windowMs: 60 * 60 * 1000 }, // 20 per hour
  'PHONE_VALIDATION': { maxRequests: 20, windowMs: 60 * 60 * 1000 }, // 20 per hour
};

export function checkRateLimit(clientId: string, operationType: string): { allowed: boolean; resetTime?: number } {
  const config = RATE_LIMIT_CONFIGS[operationType];
  if (!config) {
    return { allowed: true }; // Allow if no config found
  }
  
  const now = Date.now();
  const key = `${clientId}:${operationType}`;
  const existing = rateLimits.get(key);
  
  if (!existing || now > existing.resetTime) {
    // Create new or reset expired limit
    rateLimits.set(key, {
      count: 1,
      resetTime: now + config.windowMs
    });
    return { allowed: true };
  }
  
  if (existing.count >= config.maxRequests) {
    return {
      allowed: false,
      resetTime: existing.resetTime
    };
  }
  
  // Increment count
  existing.count++;
  rateLimits.set(key, existing);
  
  return { allowed: true };
}

export function getClientFingerprint(req: any): string {
  const ip = req.ip || req.connection.remoteAddress || 'unknown';
  const userAgent = req.get('User-Agent') || 'unknown';
  const forwarded = req.get('X-Forwarded-For') || '';
  
  // Create a simple fingerprint (in production, use more sophisticated methods)
  return `${ip}_${Buffer.from(userAgent).toString('base64').slice(0, 16)}_${forwarded}`.replace(/[^a-zA-Z0-9_]/g, '');
}

// Clean up expired rate limits periodically
setInterval(() => {
  const now = Date.now();
  for (const [key, limit] of rateLimits.entries()) {
    if (now > limit.resetTime) {
      rateLimits.delete(key);
    }
  }
}, 60 * 1000); // Clean up every minute

File: attached_assets/alberguecarcalejo-old/backend/services/validation/src/rate_limiter.rs
use crate::types::{RateLimit, RateLimitConfig};
use chrono::{DateTime, Utc};
use std::collections::HashMap;

pub struct RateLimiter {
    limits: HashMap<String, RateLimit>,
    configs: HashMap<String, RateLimitConfig>,
}

impl RateLimiter {
    pub fn new() -> Self {
        let mut configs = HashMap::new();

        configs.insert(
            "DOCUMENT_VALIDATION".to_string(),
            RateLimitConfig {
                max_requests: 10,
                window_ms: 5 * 60 * 1000, // 5 minutes
            },
        );

        configs.insert(
            "REGISTRATION".to_string(),
            RateLimitConfig {
                max_requests: 3,
                window_ms: 60 * 60 * 1000, // 1 hour
            },
        );

        configs.insert(
            "OCR_PROCESSING".to_string(),
            RateLimitConfig {
                max_requests: 5,
                window_ms: 10 * 60 * 1000, // 10 minutes
            },
        );

        configs.insert(
            "EMAIL_VALIDATION".to_string(),
            RateLimitConfig {
                max_requests: 20,
                window_ms: 60 * 60 * 1000, // 1 hour
            },
        );

        configs.insert(
            "PHONE_VALIDATION".to_string(),
            RateLimitConfig {
                max_requests: 20,
                window_ms: 60 * 60 * 1000, // 1 hour
            },
        );

        configs.insert(
            "ADMIN_AUTH".to_string(),
            RateLimitConfig {
                max_requests: 5,
                window_ms: 60 * 60 * 1000, // 1 hour
            },
        );

        configs.insert(
            "ADMIN_OPERATIONS".to_string(),
            RateLimitConfig {
                max_requests: 50,
                window_ms: 60 * 60 * 1000, // 1 hour
            },
        );

        Self {
            limits: HashMap::new(),
            configs,
        }
    }

    pub async fn check_limit(&mut self, client_id: &str, operation_type: &str) -> bool {
        let config = match self.configs.get(operation_type) {
            Some(config) => config,
            None => return true, // Allow if no config found
        };

        let now = Utc::now().timestamp_millis();
        let key = format!("{}:{}", client_id, operation_type);

        let existing = self.limits.get(&key);

        match existing {
            Some(limit) if now < limit.reset_time => {
                if limit.count >= config.max_requests {
                    false // Rate limit exceeded
                } else {
                    // Increment count
                    self.limits.insert(
                        key,
                        RateLimit {
                            count: limit.count + 1,
                            reset_time: limit.reset_time,
                        },
                    );
                    true
                }
            }
            _ => {
                // Create new or reset expired limit
                self.limits.insert(
                    key,
                    RateLimit {
                        count: 1,
                        reset_time: now + config.window_ms,
                    },
                );
                true
            }
        }
    }

    pub fn cleanup_expired(&mut self) {
        let now = Utc::now().timestamp_millis();
        self.limits.retain(|_, limit| now < limit.reset_time);
    }
}
